src/didactUriCompletionItemProvider.ts

```diff
@@ -0,0 +1,310 @@
+
+			// Add Validate All button
+			this.insertValidateAllButtonCompletion("Insert Validate All Button", completions);
+
+			// Add requirements label
+			this.insertNamedStatusLabelForRequirementCompletion("Insert Requirements Label", completions);
+
+			// Add link to install required VS Code extension
+			this.insertInstallExtensionLinkCompletion("Insert link to install required VS Code extension", completions);
+
+			// Add link to create temp folder and set it as the workspace root
+			this.insertAddWorkspaceFolderLinkCompletion("Insert link to create a temporary folder as WS root", completions);
+
+			// Add link to open didact file from elsewhere in the extension 
+			this.insertRedhatDidactLinkCompletion("Insert link to start Didact from File Elsewhere in Extension Folder", completions);
+
+			// Add didact starter 
+			this.insertDidactProtocolStarterCompletion("Start a didact link", completions);
+
+			return completions;
+		}
+		return undefined;
+	}
+
+	public provideCompletionItemsForDidactProtocol(text: string, offset: number): vscode.CompletionItem[] {
+		let completions: vscode.CompletionItem[] = [];
+		if (text.indexOf('didact://?') > -1) {
+			const start = text.indexOf('didact://?');
+			const end = text.indexOf(' ', start);
+			const parsethis = text.substring(start, end);
+			const didactUri = new DidactUri(parsethis, extContext);
+			if (didactUri) {
+				if (!didactUri.getCommandId()) {
+					completions = this.getCommandCompletionItems();
+				}
+			}
+		}
+		return completions;
+	}
+
+	private startTerminalWithNameCompletion(labelText: string, completions? : vscode.CompletionItem[]) {
+		// example: didact://?commandId=vscode.didact.startTerminalWithName&text=NamedTerminal
+		const snippetCompletion = new vscode.CompletionItem(labelText);
+		const parms = ['TerminalNameIsOptional'];
+		const snippetString = this.createCommandParmSnippetString({ commandId: extensionFunctions.START_TERMINAL_COMMAND, parms });
+		snippetCompletion.insertText = new vscode.SnippetString(snippetString);
+		snippetCompletion.documentation = new vscode.MarkdownString("Inserts a snippet that enables you to set the Terminal Name for the new Terminal.");
+		completions?.push(snippetCompletion);
+	}
+
+	private sendNamedTerminalAStringCompletion(labelText: string, completions? : vscode.CompletionItem[]) {
+		// example: didact://?commandId=vscode.didact.sendNamedTerminalAString&text=NamedTerminal$$ping%20localhost
+		const snippetCompletion = new vscode.CompletionItem(labelText);
+		const parms = ['TerminalName', 'URLEncodedTextToSendTerminal'];
+		const snippetString = this.createCommandParmSnippetString({ commandId: extensionFunctions.SEND_TERMINAL_SOME_TEXT_COMMAND, parms });
+		snippetCompletion.insertText = new vscode.SnippetString(snippetString);
+		snippetCompletion.documentation = new vscode.MarkdownString("Inserts a snippet that enables you to set the Terminal Name and Text to send the Terminal.");
+		completions?.push(snippetCompletion);
```

Comment:
is that the implicit
`if (completions) completions.push(snippetCompletion);`
?

Reply:
Yes. That's what that implies. I should make them all mandatory, not optional probably (except one case where it is optional)